streams/src/main/java/org/apache/kafka/streams/state/internals/VersionedStoreQueryUtils.java

```diff
@@ -0,0 +1,189 @@
+            result = QueryResult.notUpToBound(
+                position,
+                positionBound,
+                context == null ? null : context.taskId().partition()
+            );
+        } else {
+            result = (QueryResult<R>) handler.apply(
+                query,
+                positionBound,
+                config,
+                store
+            );
+        }
+        if (config.isCollectExecutionInfo()) {
+            result.addExecutionInfo(
+                "Handled in " + store.getClass() + " in " + (System.nanoTime() - start) + "ns"
+            );
+        }
+        result.setPosition(position);
+        return result;
+    }
+
+    public static boolean isPermitted(
+        final Position position,
+        final PositionBound positionBound,
+        final int partition
+    ) {
+        final Position bound = positionBound.position();
+        for (final String topic : bound.getTopics()) {
+            final Map<Integer, Long> partitionBounds = bound.getPartitionPositions(topic);
+            final Map<Integer, Long> seenPartitionPositions = position.getPartitionPositions(topic);
+            if (!partitionBounds.containsKey(partition)) {
+                // this topic isn't bounded for our partition, so just skip over it.
+            } else {
+                if (!seenPartitionPositions.containsKey(partition)) {
+                    // we haven't seen a partition that we have a bound for
+                    return false;
+                } else if (seenPartitionPositions.get(partition) < partitionBounds.get(partition)) {
+                    // our current position is behind the bound
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <R> QueryResult<R> runVersionedKeyQuery(final Query<R> query,
+                                                  final PositionBound positionBound,
+                                                  final QueryConfig config,
+                                                  final StateStore store) {
+        if (store instanceof VersionedKeyValueStore) {
+            final VersionedKeyValueStore<Bytes, byte[]> versionedKeyValueStore =
+                (VersionedKeyValueStore<Bytes, byte[]>) store;
+            if (query instanceof VersionedKeyQuery) {
+                final VersionedKeyQuery<Bytes, byte[]> rawKeyQuery =
+                    (VersionedKeyQuery<Bytes, byte[]>) query;
+                try {
+                    final VersionedRecord<byte[]> bytes;
+                    if (((VersionedKeyQuery<?, ?>) query).asOfTimestamp().isPresent()) {
+                        bytes = versionedKeyValueStore.get(rawKeyQuery.key(),
+                            ((VersionedKeyQuery<?, ?>) query).asOfTimestamp().get()
+                                .getLong(ChronoField.INSTANT_SECONDS));
```

Comment:
I think we need to use `toEpochMilli()` here to get the `Instant` converted to `long` correctly?